pkg/tsdb/influxdb/fsql/arrow.go

```diff
@@ -0,0 +1,277 @@
+	if err != nil {
+		resp.Error = err
+	}
+	if frame.Rows() == 0 {
+		resp.Frames = data.Frames{}
+		return resp
+	}
+
+	frame.Meta.Custom = map[string]any{
+		"headers": headers,
+	}
+	frame.Meta.ExecutedQueryString = query.RawSQL
+	frame.Meta.DataTopic = data.DataTopic(query.RawSQL)
+
+	switch query.Format {
+	case sqlutil.FormatOptionTimeSeries:
+		if _, idx := frame.FieldByName("time"); idx == -1 {
+			resp.Error = fmt.Errorf("no time column found")
+			return resp
+		}
+
+		if frame.TimeSeriesSchema().Type == data.TimeSeriesTypeLong {
+			var err error
+			frame, err = data.LongToWide(frame, nil)
+			if err != nil {
+				resp.Error = err
+				return resp
+			}
+		}
+	case sqlutil.FormatOptionTable:
+		// No changes to the output. Send it as is.
+	case sqlutil.FormatOptionLogs:
+		// TODO(brett): We need to find out what this actually is and if its
+		// worth supporting. Pass through as "table" for now.
+	default:
+		resp.Error = fmt.Errorf("unsupported format")
+	}
+
+	resp.Frames = data.Frames{frame}
+	return resp
+}
+
+// frameForRecords creates a [data.Frame] from a stream of [arrow.Record]s.
+func frameForRecords(reader recordReader) (*data.Frame, error) {
+	var (
+		frame = newFrame(reader.Schema())
+		rows  int64
+	)
+	for reader.Next() {
+		record := reader.Record()
+		for i, col := range record.Columns() {
+			if err := copyData(frame.Fields[i], col); err != nil {
+				return frame, err
+			}
+		}
+
+		rows += record.NumRows()
+		if rows > rowLimit {
+			frame.AppendNotices(data.Notice{
+				Severity: data.NoticeSeverityWarning,
+				Text:     fmt.Sprintf("Results have been limited to %v because the SQL row limit was reached", rowLimit),
+			})
+			return frame, nil
+		}
+
+		if err := reader.Err(); err != nil && !errors.Is(err, io.EOF) {
+			return frame, err
+		}
+	}
+	return frame, nil
+}
+
+// newFrame builds a new Data Frame from an Arrow Schema.
+func newFrame(schema *arrow.Schema) *data.Frame {
+	fields := schema.Fields()
+	df := &data.Frame{
+		Fields: make([]*data.Field, len(fields)),
+		Meta:   &data.FrameMeta{},
+	}
+	for i, f := range fields {
+		df.Fields[i] = newField(f)
+	}
+	return df
+}
+
+func newField(f arrow.Field) *data.Field {
+	switch f.Type.ID() {
+	case arrow.STRING:
+		return newDataField[string](f)
+	case arrow.FLOAT32:
+		return newDataField[float32](f)
+	case arrow.FLOAT64:
+		return newDataField[float64](f)
+	case arrow.UINT8:
+		return newDataField[uint8](f)
+	case arrow.UINT16:
+		return newDataField[uint16](f)
+	case arrow.UINT32:
+		return newDataField[uint32](f)
+	case arrow.UINT64:
+		return newDataField[uint64](f)
+	case arrow.INT8:
+		return newDataField[int8](f)
+	case arrow.INT16:
+		return newDataField[int16](f)
+	case arrow.INT32:
+		return newDataField[int32](f)
+	case arrow.INT64:
+		return newDataField[int64](f)
+	case arrow.BOOL:
+		return newDataField[bool](f)
+	case arrow.TIMESTAMP:
+		return newDataField[time.Time](f)
+	case arrow.DURATION:
+		return newDataField[int64](f)
+	default:
```

Comment:
There are some SQL functions such as `selector_last` that return columns of type `arrow.STRUCT`. For grafana it would probably be useful to decode the value out of these. These are structs with a `value` and `time` member, where the `value` member could be any of the valid scalar types.

It is relativly simple to decode these in the SQL however so it may be unnecessary to handle them in the plugin.